tensordict/persistent.py

```diff
@@ -0,0 +1,840 @@
+    ) -> PersistentTensorDict:
+        visitor = _Visitor()
+        self.file.visit(visitor)
+        key = self._process_key(key)
+        if key not in visitor:
+            raise KeyError(f'key "{key}" not found in h5.')
+        # we don't need to check shape as the modification will be done
+        # in-place and an error will be thrown anyway if shapes don't match
+        return self._set(key, value, inplace=True, idx=idx, check_shape=False)
+
+    def clone(self, recurse: bool = True, newfile=None) -> PersistentTensorDict:
+        if recurse:
+            # this should clone the h5 to a new location indicated by newfile
+            if newfile is None:
+                warnings.warn(
+                    "A destination should be provided when cloning a "
+                    "PersistentTensorDict. A temporary file will be used "
+                    "instead. Use `recurse=False` to keep track of the original data "
+                    "with a new PersistentTensorDict instance."
+                )
+                tmpfile = tempfile.NamedTemporaryFile()
+                newfile = tmpfile.name
+            f_dest = h5py.File(newfile, "w")
+            f_src = self.file
+            for key in self.keys(True, True):
+                key = self._process_key(key)
+                f_dest.create_dataset(key, data=f_src[key], **self.kwargs)
+                # f_src.copy(f_src[key],  f_dest[key], "DataSet")
+            # create a non-recursive copy and update the file
+            # this way, we can keep the batch-size of every nested tensordict
+            clone = self.clone(False)
+            clone.file = f_src
+            clone.filename = newfile
+            clone._pin_mem = False
+            return clone
+        else:
+            # we need to keep the batch-size of nested tds, which we do manually
+            nested_tds = {
+                key: td.clone(False) for key, td in self._nested_tensordicts.items()
+            }
+            filename = self.filename
+            file = self.file if filename is None else None
+            clone = PersistentTensorDict(
+                filename=filename,
+                group=file,
+                mode=self.mode,
+                backend="h5",
+                device=self.device,
+                batch_size=self.batch_size,
+            )
+            clone._nested_tensordicts = nested_tds
+            clone._pin_mem = False
+            return clone
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        filename = state["file"].file.filename
+        group_name = state["file"].name
+        state["file"] = None
+        state["filename"] = filename
+        state["group_name"] = group_name
+        return state
+
+    def __setstate__(self, state):
+        state["file"] = h5py.File(state["filename"], mode=state["mode"])
+        if state["group_name"] != "/":
+            state["file"] = state["file"][state["group_name"]]
+        del state["group_name"]
```

Comment:
Why delete `group_name` but not `filename`?

Reply:
because we don't have an attribute groupname in the class
but it's not super elegant I admit